pydantic/json_schema.py

```diff
@@ -1658,3 +1660,59 @@ def _sort_json_schema(value: JsonSchemaValue) -> JsonSchemaValue:
         return sorted_list  # type: ignore
     else:
         return value
+
+
+@_internal_dataclass.slots_dataclass
+class WithJsonSchema:
+    """Add this as an annotation on a field to override the (base) JSON schema that would be generated for that field.
+    This provides a way to set a JSON schema for types that would otherwise raise errors when producing a JSON schema,
+    such as Callable, or types that have an is-instance core schema, without needing to go so far as creating a
+    custom subclass of pydantic.json_schema.GenerateJsonSchema.
+    Note that any _modifications_ to the schema that would normally be made (such as setting the title for model fields)
+    will still be performed.
+
+    If `mode` is set this will only apply to that schema generation mode, allowing you
+    to set different json schemas for validation and serialization.
+    """
+
+    json_schema: JsonSchemaValue | None
+    mode: Literal['validation', 'serialization'] | None = None
+
+    def __get_pydantic_json_schema__(
+        self, core_schema: core_schema.CoreSchema, handler: _annotated_handlers.GetJsonSchemaHandler
+    ) -> JsonSchemaValue:
+        mode = self.mode or handler.mode
+        if mode != handler.mode:
+            return handler(core_schema)
+        if self.json_schema is None:
+            # This exception is handled in pydantic.json_schema.GenerateJsonSchema._named_required_fields_schema
+            raise PydanticOmit
+        else:
+            return self.json_schema
+
+
+@_internal_dataclass.slots_dataclass
+class Examples:
+    """Add examples to a JSON schema.
+
+    Examples should be a map of example names (strings)
+    to example values (any valid JSON).
+
+    If `mode` is set this will only apply to that schema generation mode,
+    allowing you to add different examples for validation and serialization.
+    """
```

Comment:
@tpdorsey any suggestions as to how to add docs/examples for these things? I see that they should be in the API docs but I'm not sure if I can put examples in there or if I need to make a new markdown block/section, etc.

Reply:
@adriangb check field.py and functional_validators.py for examples of examples in docstrings.

You don't have to use the "Example:" heading. That just puts the example in a pop-out admonition when built in the docs. You can also include the example code inline with the description part of the docstring.

Note that examples are tested even in docstrings.

Reply:
I'm pretty sure about that tested part because I recall getting hung up on it recently.

Reply:
Okay I see: https://github.com/pydantic/pydantic/blob/beff6468d22c38adb950fed430db4066045e4272/pydantic/functional_validators.py#L31-L63

Reply:
Bad news: these examples are not run. I put an `assert False` in one and it didn't work.
They also don't get formatted, type-checked, etc.